GraalVM Native Image failed to load the required native library

The jar file works fine but the native image compiled by GraalVM (21.0.2+13.1) throws UnsatisfiedLinkError exception:

    java.lang.UnsatisfiedLinkError: failed to load the required native library
            at io.netty.incubator.codec.quic.Quic.ensureAvailability(Quic.java:80)
            at io.netty.incubator.codec.quic.QuicheQuicSslContext.<init>(QuicheQuicSslContext.java:78)
            at io.netty.incubator.codec.quic.QuicSslContextBuilder.build(QuicSslContextBuilder.java:371)
            at common.net.QuicStreamChannelHandler.listen(QuicStreamChannelHandler.java:306)
            at common.net.ServerEndpoint.listen(ServerEndpoint.java:85)
            at myapp.AgencyMgr.login(AgencyMgr.java:279)
            at myapp.ui.UIMgr.lambda$new$2(UIMgr.java:26)
            at myapp.ui.Tui$1.run(Tui.java:93)
            at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:833)
            at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:211)
    Caused by: java.lang.UnsatisfiedLinkError: Can't load library: /tmp/libnetty_quiche_linux_x86_6413187606951079597579.so
            at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.NativeLibrarySupport.loadLibraryAbsolute(NativeLibrarySupport.java:100)
            at java.base@21.0.2/java.lang.ClassLoader.loadLibrary(ClassLoader.java:114)
            at java.base@21.0.2/java.lang.Runtime.load0(Runtime.java:852)
            at java.base@21.0.2/java.lang.System.load(System.java:2025)
            at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:36)
            at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:396)
            at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:218)
            at io.netty.incubator.codec.quic.Quiche.loadNativeLibrary(Quiche.java:77)
            at io.netty.incubator.codec.quic.Quiche.<clinit>(Quiche.java:57)
            at io.netty.incubator.codec.quic.Quic.<clinit>(Quic.java:45)
            ... 9 more
            Suppressed: java.lang.UnsatisfiedLinkError: Can't load library: /tmp/libnetty_quiche_linux_x86_6413187606951079597579.so
                    at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.NativeLibrarySupport.loadLibraryAbsolute(NativeLibrarySupport.java:100)
                    at java.base@21.0.2/java.lang.ClassLoader.loadLibrary(ClassLoader.java:114)
                    at java.base@21.0.2/java.lang.Runtime.load0(Runtime.java:852)
                    at java.base@21.0.2/java.lang.System.load(System.java:2025)
                    at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:36)
                    at java.base@21.0.2/java.lang.reflect.Method.invoke(Method.java:580)
                    at io.netty.util.internal.NativeLibraryLoader$1.run(NativeLibraryLoader.java:430)
                    at java.base@21.0.2/java.security.AccessController.executePrivileged(AccessController.java:129)
                    at java.base@21.0.2/java.security.AccessController.doPrivileged(AccessController.java:319)
                    at io.netty.util.internal.NativeLibraryLoader.loadLibraryByHelper(NativeLibraryLoader.java:422)
                    at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:388)
                    ... 13 more
            Suppressed: java.lang.UnsatisfiedLinkError: No netty_quiche_linux_x86_64 in java.library.path
                    at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.NativeLibrarySupport.loadLibraryRelative(NativeLibrarySupport.java:136)
                    at java.base@21.0.2/java.lang.ClassLoader.loadLibrary(ClassLoader.java:106)
                    at java.base@21.0.2/java.lang.Runtime.loadLibrary0(Runtime.java:916)
                    at java.base@21.0.2/java.lang.System.loadLibrary(System.java:2063)
                    at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:38)
                    at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:396)
                    at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:161)
                    ... 12 more
                    Suppressed: java.lang.UnsatisfiedLinkError: No netty_quiche_linux_x86_64 in java.library.path
                            at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.NativeLibrarySupport.loadLibraryRelative(NativeLibrarySupport.java:136)
                            at java.base@21.0.2/java.lang.ClassLoader.loadLibrary(ClassLoader.java:106)
                            at java.base@21.0.2/java.lang.Runtime.loadLibrary0(Runtime.java:916)
                            at java.base@21.0.2/java.lang.System.loadLibrary(System.java:2063)
                            at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:38)
                            at java.base@21.0.2/java.lang.reflect.Method.invoke(Method.java:580)
                            at io.netty.util.internal.NativeLibraryLoader$1.run(NativeLibraryLoader.java:430)
                            at java.base@21.0.2/java.security.AccessController.executePrivileged(AccessController.java:129)
                            at java.base@21.0.2/java.security.AccessController.doPrivileged(AccessController.java:319)
                            at io.netty.util.internal.NativeLibraryLoader.loadLibraryByHelper(NativeLibraryLoader.java:422)
                            at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:388)
                            ... 13 more

Here is the code snippet in my `QuicStreamChannelHandler.java`:

        QuicSslContext sslCtx = QuicSslContextBuilder.forClient()
                .trustManager(cert)
                .applicationProtocols("http/1.1").build();
        ChannelHandler codec = new QuicClientCodecBuilder().sslContext(sslCtx)
                .maxIdleTimeout(2000, TimeUnit.MILLISECONDS)
                .initialMaxData(23450)
                .initialMaxStreamDataBidirectionalLocal(23450)
                .build();

In the jar file, `META-INF/native-image/io.netty.incubator/netty-incubator-codec-native-quic/resource-config.json` is there with the contents:

    {
      "resources":{
      "includes":[
        {
          "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
          "pattern":"\\QMETA-INF/native/libnetty_quiche_osx_x86_64.jnilib\\E"
        },
        {
          "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
          "pattern":"\\QMETA-INF/native/libnetty_quiche_osx_aarch_64.jnilib\\E"
        },
        {
          "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
          "pattern":"\\QMETA-INF/native/libnetty_quiche_linux_x86_64.so\\E"
        },
        {
          "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
          "pattern":"\\QMETA-INF/native/libnetty_quiche_linux_aarch_64.so\\E"
        },
        {
          "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
          "pattern":"\\QMETA-INF/native/netty_quiche_windows_x86_64.dll\\E"
        }
      ]},
      "bundles":[]
    }

and in META-INF/native/, file `libnetty_quiche_linux_x86_64.so` and `netty_quiche_windows_x86_64.dll` are there.
I noticed that in the exception it says can't load library */tmp/libnetty_quiche_linux_x86_64**13187606951079597579**.so*, I didn't dig into `Quic.ensureAvailability` and don't understand where the value *13187606951079597579* (looks like a random number) came from, and how `/tmp/libnetty_quiche_linux_x86_6413187606951079597579.so` maps to `META-INF/native/libnetty_quiche_linux_x86_64.so`.

--- a/codec-native-quic/src/main/resources/META-INF/native-image/io.netty.incubator/netty-incubator-codec-native-quic/jni-config.json
+++ b/codec-native-quic/src/main/resources/META-INF/native-image/io.netty.incubator/netty-incubator-codec-native-quic/jni-config.json
@@ -105,7 +105,7 @@
 {
   "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
   "name":"java.lang.Integer",
-  "methods":[{"name":"<init>","parameterTypes":["int"] }]
+  "methods":[{"name":"<init>","parameterTypes":["int"] },{"name":"valueOf","parameterTypes":["int"]} ]
 },
 {
   "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
@@ -114,5 +114,34 @@
 {
   "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
   "name":"java.lang.String"
+},
+{
+  "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
+  "name":"io.netty.incubator.codec.quic.BoringSSLSessionTicketCallback",
+  "methods":[{"name":"findSessionTicket","parameterTypes":["byte[]"] }]
+},
+{
+  "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
+  "name":"java.lang.Long",
+  "methods":[{"name":"valueOf","parameterTypes":["long"] }]
+},
+{
+  "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
+  "name":"java.net.Inet4Address"
+},
+{
+  "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
+  "name":"java.net.Inet6Address",
+  "methods":[{"name":"getByAddress","parameterTypes":["java.lang.String","byte[]","int"] }]
+},
+{
+  "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
+  "name":"java.net.InetAddress",
+  "methods":[{"name":"getByAddress","parameterTypes":["byte[]"] }]
+},
+{
+  "condition":{"typeReachable":"io.netty.incubator.codec.quic.Quiche"},
+  "name":"java.net.InetSocketAddress",
+  "methods":[{"name":"<init>","parameterTypes":["java.net.InetAddress","int"] }]
 }
 ]

